Itertools.php

```diff
@@ -0,0 +1,258 @@
+   * start is non-zero, then elements from the iterable are skipped until start
+   * is reached. Afterward, elements are returned consecutively unless step is
+   * set higher than one which results in items being skipped. If stop is not
+   * set, then iteration continues until the iterator is exhausted, if at all;
+   * otherwise, it stops at the specified position. Unlike regular slicing,
+   * islice() does not support negative values for start, stop, or step.
+  */
+  function islice($iterable, ...$args) {
+    return iter(function() use ($iterable, $args) {
+      $s = new Slice(...$args);
+      $indices = iter(range($s->start, $s->stop-1, $s->step));
+
+      try {
+        $next_index = inext($indices);
+      } catch (StopIteration $e) {
+        return;
+      }
+
+      foreach (enumerate($iterable) as $index => $element) {
+        if ($index === $next_index) {
+          yield $element;
+          try {
+            $next_index = inext($indices);
+          } catch (StopIteration $e) {
+            return;
+          }
+        }
+      }
+    });
+  }
+
+  /**
+   * Make an iterator returning elements from the iterable and saving a copy of
+   * each. When the iterable is exhausted, return elements from the saved copy.
+   * Repeats indefinitely.
+  */
+  function cycle($iterable) {
+    return iter(function() use ($iterable) {
+      $saved = new Arr();
+      foreach ($iterable as $element) {
+        yield $element;
+        $saved->append($element);
+      }
+      while (count($saved) > 0) {
+        foreach ($saved as $element) {
+          yield $element;
+        }
+      }
+    });
+  }
+
+  // TODO: accumulate -> maybe a method on Genewrapor?
+
+  /**
+   * Make an iterator that returns elements from the first iterable until it is
+   * exhausted, then proceeds to the next iterable, until all of the iterables
+   * are exhausted. Used for treating consecutive sequences as a single
+   * sequence.
+  */
+  function chain(...$iterables) {
+    return iter(function() use ($iterables) {
+      foreach ($iterables as $it) {
+        foreach ($it as $element) {
+          yield $element;
+        }
+      }
+    });
+  }
+
+  // TODO: groupby is cool, but should maybe be a method of dict
+  // TODO: i think I want some `join` function. implode doesn't work on Genewrapors
+
+  /**
+   * Return successive r length permutations of elements in the iterable.
+   *
+   * If r is not specified or is None, then r defaults to the length of the
+   * iterable and all possible full-length permutations are generated.
+   *
+   * Permutations are emitted in lexicographic sort order. So, if the input
+   * iterable is sorted, the permutation tuples will be produced in sorted
+   * order.
+   *
+   * Elements are treated as unique based on their position, not on their
+   * value. So if the input elements are unique, there will be no repeat values
+   * in each permutation.
+  */
+  function permutations($iterable, $r=null) {
+
```

Comment:
please do either:
- add TODO
- add implementation
-  uncomment the whole function and add issue (feaure) after having merged the pull request